### PR TITLE
Triple indexed syntax

### PR DESCRIPTION
**Pull request by miaoyinb**
ref. #21930

### CODE REVIEW

**Comment by miaoyinb** [comment]
Just a draft for now. Need to wait the libmesh PR and at least add a test object.

File: framework/src/parser/Parser.C [hunk 1]
```diff
@@ -1848,6 +1896,53 @@ Parser::setDoubleIndexParameter(const std::string & full_name,
...
+    GlobalParamsAction * global_block)
+{
+  // Get the full string assigned to the variable full_name
+  std::string buffer = _root->param<std::string>(full_name);
+
+  // split vector at delim ;
+  // NOTE: the substrings are _not_ of type T yet
+  std::vector<std::string> first_tokenized_vector;
+  std::vector<std::vector<std::string>> second_tokenized_vector;
+  MooseUtils::tokenize(buffer, first_tokenized_vector, 1, "|");
```
Comment (miaoyinb) [inline framework/src/parser/Parser.C:1915]: I am using "|" for now but this is definitely open for discussion.

**Comment by miaoyinb** [comment]
@GiudGiud @loganharbour 

The libMesh PR has been merged (https://github.com/libMesh/libmesh/issues/3380). Please advise how we should proceed here. Thanks.

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: test/src/userobjects/ReadTripleIndex.C [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+      mooseError("Error reading uint_tri.");
+    for (unsigned int j = 0; j < _uint_tri[i].size(); j++)
+    {
+      if (_uint_tri[i][j].size() != reference_values[i][j].size())
+        mooseError("Error reading uint_tri.");
+      for (unsigned int k = 0; k < _uint_tri[i][j].size(); k++)
+        if (_uint_tri[i][j][k] != static_cast<unsigned int>(reference_values[i][j][k] * 10.0))
+          mooseError("Error reading uint_tri.");
+    }
+  }
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply test/src/userobjects/ReadTripleIndex.C:126]: Will do, I was too lazy.

**Comment by dschwen** [comment]
I'm not convinced this is a good idea. 3d arrays w will be large and unwieldy and not suitable for input parameters

**Comment by miaoyinb** [comment]
> I'm not convinced this is a good idea. 3d arrays w will be large and unwieldy and not suitable for input parameters

We do have several cases that 3D arrays could make the input syntax clearer, including the element extra integer swap parameter in`FancyExtruderGenerator`.

**Comment by GiudGiud** [comment]
The extrusion cases they have are indexed by axial plane, then hold both an origin and a target number. 
We re open to suggestions but this works imo

**Comment by YaqiWang** [comment]
> The extrusion cases they have are indexed by axial plane, then hold both an origin and a target number. We re open to suggestions but this works imo

Sometimes using map type for parameters can make the input more straightforward. If you have your own parsing function as I suggested in the issue, you can add whatever keywords to make the syntax more readable like
```
layer 1
x y z

layer 0
a b c
```
where `layer` can serve as a keyword. It is hard to be convincing that the 3-d vector type is needed.

**Comment by miaoyinb** [comment]
> > The extrusion cases they have are indexed by axial plane, then hold both an origin and a target number. We re open to suggestions but this works imo
> 
> Sometimes using map type for parameters can make the input more straightforward. If you have your own parsing function as I suggested in the issue, you can add whatever keywords to make the syntax more readable like
> 
> ```
> layer 1
> x y z
> 
> layer 0
> a b c
> ```
> 
> where `layer` can serve as a keyword. It is hard to be convincing that the 3-d vector type is needed.

Actually,  in `FancyExtruderGenerator`, we need the triple-indexed vector form for the `elem_integers_swaps` parameter.
The idea originate from the subdomain swap idea. For example. if we have subdomain ids 1, 2, and 3 (e.g., fuel, insulation, reflector) in the original 2D mesh and want to extrude it into three layers of elevation. After extrusion, we would like to make all top and bottom layer to be subdomain 3 (reflector). To do so, we can set 

```
subdomain_swaps = '1 3 2 3 3 3;
                   1 1 2 2 3 3;
                   1 3 2 3 3 3'
```
or the simple form without unchanged pairs.

```
subdomain_swaps = '1 3 2 3 ;
                   ;
                   1 3 2 3'
```

Here `subdomain_swaps` is a 2D vector, the first dimension is extruded layer (i.e., elevation). the second dimension is pairs of subdomain ids that need to be swapped (the first is the original, the second is the wanted value).

To use the similar approach for element extra integers such as `material_id` and `deletion_id`. The major challenge is that we may have more than one set of element extra integers that need to be swapped. Currently, the sets that need to be swapped are defined by `elem_integer_names_to_swap`; and then the swap rules of multiple sets of integers are stacked in `elem_integers_swaps`. For example,

```
    elem_integer_names_to_swap = 'element_extra_integer_1 element_extra_integer_2'
    elem_integers_swaps = '1 4 2 8;
                           2 7;
                           1 6;      #first three lines are for element_extra_integer_1 
                           1 8 2 4;
                           2 5;
                           1 6'      #last three lines are for element_extra_integer_2
```
We think using a 3D vector instead of stacking them into a 2D vector would be helpful. Like

```
    elem_integer_names_to_swap = 'element_extra_integer_1 element_extra_integer_2'
    elem_integers_swaps = '1 4 2 8;
                           2 7;
                           1 6|      # "|" is used here for the new dimension
                           1 8 2 4;
                           2 5;
                           1 6'    
```

**Comment by dschwen** [comment]
Three dimensional vectors are going to be unintuitive. Both one and two D can easily be laid out in the input. The third dimension is not easy to lay out intuitively in a 2D text file.

**Comment by miaoyinb** [comment]
> Three dimensional vectors are going to be unintuitive. Both one and two D can easily be laid out in the input. The third dimension is not easy to lay out intuitively in a 2D text file.

This is true as we are all limited by 2D screens. 

Here we just want to have the third type of numeric separators (aside from space and semicolon) to make some input parameters clearer.

**Review by GiudGiud — COMMENTED** [review]
3D-like (eg currently flattened in one axis) vectors are not new in MOOSE parameters and are less clear than true 3D vector input. 
The extrusion code has long had:
- 2D plane data, with an outer 1D axial indexing
- swapped pairs, with an outer 1D axial indexing

custom map parsing would on the contrary be completely new. Long term we need to think:
- what is easiest to QA?
- what is easiest to parse automatically for plugin-laguage-server assisted inputs?

**Comment by YaqiWang** [comment]
> 

The input can be made into
```
    elem_integers_swaps = '
                          element_extra_integer_1
                           1 4 2 8;
                           2 7;
                           1 6

                           element_extra_integer_2
                           1 8 2 4;
                           2 5;
                           1 6' 
```
for example, with a custom parsing function. It can be even made into something like
```
    elem_integers_swaps = '
                          element_extra_integer_1
                          layer 0 1=>4 2=>8
                          layer 1 2=>7
                          layer 2 1=>6

                           element_extra_integer_2
                           layer 0 1=>8 2=>4
                           layer 1 2=>5
                           layer 2 1=>6' 
```

I do not have a super strong opinion on 3D vectors in inputs but I agree with @dschwen that they are very counter intuitive.

**Comment by GiudGiud** [comment]
We could include the layer indexing or the extra_integer name as a comment and achieve the same. These two examples you are showing are 3D vectors, just with explicit declaration of the indexing on one or two indexes.

This custom parsing function is wishful thinking at this point. It would introduce as many difficulties as it would solve. i dont see it working well at all with the language-server work. 

3D is not **counter** intuitive. It's maybe not intuitive but it definitely does not defy or counter logic. It's in direct continuation of systems that are already in MOOSE.

**Comment by YaqiWang** [comment]
> We could include the layer indexing or the extra_integer name as a comment and achieve the same. These two examples you are showing are 3D vectors, just with explicit declaration of the indexing on one or two indexes.
> 
> This custom parsing function is wishful thinking at this point. It would introduce as many difficulties as it would solve. i dont see it working well at all with the language-server work.
> 
> 3D is not **counter** intuitive. It's maybe not intuitive but it definitely does not defy or counter logic. It's in direct continuation of systems that are already in MOOSE.

The reason you were giving is not user oriented. As I said, if you insist, I am ok.

File: test/tests/parser/parse_triple_index/parse_triple_index.i [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  solve_type = 'PJFNK'
+
+
+  petsc_options_iname = '-pc_type -pc_hypre_type'
+  petsc_options_value = 'hypre boomeramg'
+[]
+
+[Outputs]
+  exodus = true
```
Comment (GiudGiud) [inline test/tests/parser/parse_triple_index/parse_triple_index.i:81]: no need for that, in fact no need for a problem solve either I think

**Review by GiudGiud — CHANGES_REQUESTED** [review]
the parsing seems to be totally fine with double symbols ;; and || even though the dimension should be off??

;;;;;;;;;; is parsed incorrectly too. This kind of short syntax could be used by users who want to skip a plane for example in the extrusion

**Comment by miaoyinb** [comment]
> the parsing seems to be totally fine with double symbols ;; and || even though the dimension should be off??
> 
> ;;;;;;;;;; is parsed incorrectly too. This kind of short syntax could be used by users who want to skip a plane for example in the extrusion

Good point. I think I fixed this in the commit I just pushed and added some test cases.

**Comment by miaoyinb** [comment]
@GiudGiud Finally, conda has been updated so that this one can pass all the checks.

**Review by GiudGiud — APPROVED** [review]
looks great!

Thank you for the contribution and addressing the reviews promptly and extensively!

**Pull request merged**